File: 08-Singleton/src/go/config_manager.go
package main

import (
	"fmt"
	"sync"
)

type ConfigManager struct {
	properties map[string]string
}

var (
	configInstance *ConfigManager
	configOnce     sync.Once
)

func GetConfigInstance() *ConfigManager {
	configOnce.Do(func() {
		configInstance = &ConfigManager{
			properties: make(map[string]string),
		}
		configInstance.loadDefaults()
	})
	return configInstance
}

func (cm *ConfigManager) loadDefaults() {
	cm.properties["app.name"] = "LLD Learning App"
	cm.properties["app.version"] = "1.0"
	cm.properties["db.host"] = "localhost"
	cm.properties["db.port"] = "5432"
	cm.properties["log.level"] = "INFO"
}

func (cm *ConfigManager) Get(key string) string {
	return cm.properties[key]
}

func (cm *ConfigManager) Set(key, value string) {
	cm.properties[key] = value
}

func (cm *ConfigManager) Size() int {
	return len(cm.properties)
}

func (cm ConfigManager) String() string {
	return fmt.Sprintf("ConfigManager[%d properties]", len(cm.properties))
}

File: 08-Singleton/src/rust/config_manager.rs
use std::collections::HashMap;
use std::sync::OnceLock;

pub struct ConfigManager {
    properties: HashMap<String, String>,
}

static CONFIG_INSTANCE: OnceLock<std::sync::Mutex<ConfigManager>> = OnceLock::new();

impl ConfigManager {
    fn new() -> Self {
        let mut cm = ConfigManager {
            properties: HashMap::new(),
        };
        cm.load_defaults();
        cm
    }

    fn load_defaults(&mut self) {
        self.properties.insert("app.name".into(), "LLD Learning App".into());
        self.properties.insert("app.version".into(), "1.0".into());
        self.properties.insert("db.host".into(), "localhost".into());
        self.properties.insert("db.port".into(), "5432".into());
        self.properties.insert("log.level".into(), "INFO".into());
    }

    pub fn get_instance() -> &'static std::sync::Mutex<ConfigManager> {
        CONFIG_INSTANCE.get_or_init(|| std::sync::Mutex::new(ConfigManager::new()))
    }

    pub fn get(&self, key: &str) -> Option<&String> {
        self.properties.get(key)
    }

    pub fn set(&mut self, key: &str, value: &str) {
        self.properties.insert(key.into(), value.into());
    }

    pub fn size(&self) -> usize {
        self.properties.len()
    }
}
